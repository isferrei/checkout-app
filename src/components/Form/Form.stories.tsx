import type { Meta, StoryObj } from "@storybook/react";

import { Form } from "./Form";

const mockInstallments = [
  { label: "1", value: 1 },
  { label: "2", value: 2 },
  { label: "3", value: 3 },
];

const meta = {
  title: "Example/Form",
  component: Form,
  tags: ["autodocs"],
  parameters: {
    layout: "fullscreen",
  },
} satisfies Meta;

export default meta;
type Story = StoryObj;

export const Default = {
  args: {
    installments: mockInstallments,
    onSubmit: () => {},
  },
} satisfies Story;
