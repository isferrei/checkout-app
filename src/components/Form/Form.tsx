import { FormEvent, useState } from "react";
import { Input } from "../Input/Input";
import { Button } from "../Button/Button";
import { Select } from "../Select/Select";

type FormData = {
  [key: string]: string;
};

type FormProps = {
  onSubmit: (data: FormData) => void;
  installments: Array<{
    label: string | number;
    value: string | number;
  }>;
};

export function Form(props: FormProps) {
  const [formData, setFormData] = useState({});

  const handleInputChange = (name: string, value: string) => {
    setFormData((prevData) => ({
      ...prevData,
      [name]: value,
    }));
  };

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    props.onSubmit(formData);
  };

  return (
    <form
      onSubmit={handleSubmit}
      className="flex flex-col gap-[1.87rem]"
      data-testid="form"
    >
      <Input
        label="Número do cartão"
        mask="9999 9999 9999 9999"
        placeholder="0000 0000 0000 0000"
        onChange={() => {}}
        name="cardNumber"
      />
      <div className="flex flex-row gap-[3.12rem]">
        <Input
          label="Validade"
          mask="99/99"
          placeholder="MM/AA"
          onChange={() => {}}
          name="cardValidate"
        />
        <Input
          label="CVV"
          mask="999"
          placeholder="000"
          onChange={() => {}}
          name="cardCvv"
        />
      </div>
      <Input
        label="Nome impresso no cartão"
        mask=""
        placeholder="Seu nome"
        onChange={() => {}}
        name="name"
      />
      <Input
        label="CPF"
        mask="999.999.999-99"
        placeholder="000.000.000-00"
        onChange={() => {}}
        name="cpf"
      />
      <Input
        label="Cupom"
        mask=""
        type="text"
        placeholder="Insira aqui"
        onChange={() => {}}
        name="cupom"
      />
      <Select
        label="Número de parcelas"
        options={props.installments}
        placeholder={"Selecionar"}
        onSelect={() => {}}
      />

      <div className="mt-[0.62rem]">
        <Button variation="primary">Finalizar pagamento</Button>
      </div>
    </form>
  );
}
