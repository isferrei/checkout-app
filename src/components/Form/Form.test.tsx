import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import { Form } from "./Form";

describe("Form", () => {
  const mockInstallments = [
    { label: "Option 1", value: 1 },
    { label: "Option 2", value: 2 },
    { label: "Option 3", value: 3 },
  ];

  test("calls onSubmit handler with form data on submission", () => {
    const mockOnSubmit = jest.fn();
    render(<Form onSubmit={mockOnSubmit} installments={mockInstallments} />);

    act(() => {
      userEvent.type(
        screen.getByLabelText(/Número do cartão/i),
        "1234567890123456"
      );
      userEvent.type(screen.getByLabelText(/Validade/i), "12/23");
      userEvent.type(screen.getByLabelText(/CVV/i), "123");
      userEvent.type(
        screen.getByLabelText(/Nome impresso no cartão/i),
        "John Doe"
      );
      userEvent.type(screen.getByLabelText(/CPF/i), "123.456.789-00");
      userEvent.type(screen.getByLabelText(/Cupom/i), "ABC123");

      userEvent.click(screen.getByText(/Finalizar pagamento/i));
    });

    expect(mockOnSubmit).toHaveBeenCalledTimes(1);

    expect(screen.getByLabelText(/Número do cartão/i)).toMatchSnapshot();
    expect(screen.getByLabelText(/Validade/i)).toMatchSnapshot();
    expect(screen.getByLabelText(/CVV/i)).toMatchSnapshot();
    expect(screen.getByLabelText(/Nome impresso no cartão/i)).toMatchSnapshot();
    expect(screen.getByLabelText(/CPF/i)).toMatchSnapshot();
    expect(screen.getByLabelText(/Cupom/i)).toMatchSnapshot();
    expect(screen.getByText(/Finalizar pagamento/i)).toMatchSnapshot();

    expect(screen.getByLabelText(/Número do cartão/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Validade/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/CVV/i)).toBeInTheDocument();
    expect(
      screen.getByLabelText(/Nome impresso no cartão/i)
    ).toBeInTheDocument();
    expect(screen.getByLabelText(/CPF/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Cupom/i)).toBeInTheDocument();
    expect(screen.getByText(/Finalizar pagamento/i)).toBeInTheDocument();
  });
});
