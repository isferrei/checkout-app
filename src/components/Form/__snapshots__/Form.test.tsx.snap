// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form calls onSubmit handler with form data on submission 1`] = `
<input
  class="w-full border-b-[1px] text-[1rem] text-darkGray font-sans border-[#F4F3F6] py-[0.25rem] pl-1 placeholder:text-lightGray focus:outline-none focus:ring-2"
  id="creditCardNumber"
  name="creditCardNumber"
  placeholder="0000 0000 0000 0000"
  required=""
  type="text"
  value=""
/>
`;

exports[`Form calls onSubmit handler with form data on submission 2`] = `
<input
  class="w-full border-b-[1px] text-[1rem] text-darkGray font-sans border-[#F4F3F6] py-[0.25rem] pl-1 placeholder:text-lightGray focus:outline-none focus:ring-2"
  id="creditCardExpirationDate"
  name="creditCardExpirationDate"
  placeholder="MM/AA"
  required=""
  type="text"
  value=""
/>
`;

exports[`Form calls onSubmit handler with form data on submission 3`] = `
<input
  class="w-full border-b-[1px] text-[1rem] text-darkGray font-sans border-[#F4F3F6] py-[0.25rem] pl-1 placeholder:text-lightGray focus:outline-none focus:ring-2"
  id="creditCardCVV"
  name="creditCardCVV"
  placeholder="000"
  required=""
  type="text"
  value=""
/>
`;

exports[`Form calls onSubmit handler with form data on submission 4`] = `
<input
  class="w-full border-b-[1px] text-[1rem] text-darkGray font-sans border-[#F4F3F6] py-[0.25rem] pl-1 placeholder:text-lightGray focus:outline-none focus:ring-2"
  id="creditCardHolder"
  name="creditCardHolder"
  placeholder="Seu nome"
  required=""
  type="text"
  value=""
/>
`;

exports[`Form calls onSubmit handler with form data on submission 5`] = `
<input
  class="w-full border-b-[1px] text-[1rem] text-darkGray font-sans border-[#F4F3F6] py-[0.25rem] pl-1 placeholder:text-lightGray focus:outline-none focus:ring-2"
  id="creditCardCPF"
  name="creditCardCPF"
  placeholder="000.000.000-00"
  required=""
  type="text"
  value=""
/>
`;

exports[`Form calls onSubmit handler with form data on submission 6`] = `
<input
  class="w-full border-b-[1px] text-[1rem] text-darkGray font-sans border-[#F4F3F6] py-[0.25rem] pl-1 placeholder:text-lightGray focus:outline-none focus:ring-2"
  id="couponCode"
  name="couponCode"
  placeholder="Insira aqui"
  required=""
  type="text"
  value=""
/>
`;

exports[`Form calls onSubmit handler with form data on submission 7`] = `
<button
  class="flex justify-center items-center w-full rounded-[25px] text-[0.94rem] bg-blue text-white disabled:bg-gray hover:bg-lightBlue transition-all"
  style="height: 50px; font-weight: 400;"
  type="submit"
>
  Finalizar pagamento
</button>
`;
